partiql-parser/src/lalr/lexer.rs

```diff
@@ -1,7 +1,64 @@
+use crate::result::LineAndColumn;
+use itertools::Itertools;
 use logos::{Logos, Span};
+use smallvec::SmallVec;
 use std::cmp::max;
+use std::ops::Range;
 use thiserror::Error;
 
+/// Keeps track of source offsets of newlines for the purposes of later calculating
+/// line and column information
+///
+///
+/// ## Example
+///
+/// ```rust
+/// use partiql_parser::result::LineAndColumn;
+/// use partiql_parser::LineOffsetTracker;
+///
+/// let mut tracker = LineOffsetTracker::default();
+/// tracker.record(5..6);
+/// tracker.record(15..16);
+/// tracker.record(25..26);
+///
+/// // We added 3 newlines, so there should be 4 lines of source
+/// assert_eq!(tracker.lines(), 4);
+/// assert_eq!(tracker.at(0), LineAndColumn::at(1,1));
+/// assert_eq!(tracker.at(6), LineAndColumn::at(2,1));
+/// assert_eq!(tracker.at(400), LineAndColumn::at(4,375));
+/// ```
+#[derive(Default)]
+pub struct LineOffsetTracker {
+    newlines: SmallVec<[usize; 8]>,
+}
+
+impl LineOffsetTracker {
```

Comment:
Revisiting this, I realize this won't handle unicode. We'll need to actually count the chars in a line, not just convert the byte offsets into column offsets.